design/open/openapi-multipart-upload.md

```diff
@@ -0,0 +1,215 @@
+        required: true
+        schema:
+          type: string
+    put:
+      tags:
+        - staging
+      operationId: completeMultipartUpload
+      summary: Complete a multipart upload
+      description: Completes a multipart upload by assembling the uploaded parts.
+      requestBody:
+        content:
+          application/json:
+            schema:
+              $ref: "#/components/schemas/CompleteMultipartUpload"
+      responses:
+        200:
+          description: Multipart upload completed
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/ObjectStats"
+        # ... other error responses
+
+    delete:
+      tags:
+        - staging
+      operationId: abortMultipartUpload
+      summary: Abort a multipart upload
+      description: Aborts a multipart upload.
+      responses:
+        204:
+          description: Multipart upload aborted
+
+```
+
+
+**Schemas used by operations**
+
+```yaml
+components:
+  schemas:
+    MultipartUpload:
+      type: object
+      properties:
+        upload_id:
+          type: string
+        locations:
+          type: array
+          items:
+            $ref: "#/components/schemas/StagingLocation"
+      required:
+        - upload_id
+        
+    UploadPart:
+      type: object
+      properties:
+        part_number:
+          type: integer
+        etag:
+          type: string
+      required:
+        - part_number
+        - etag
+        
+    CompleteMultipartUpload:
+      type: object
+      properties:
+        parts:
+          type: array
+          description: "List of uploaded parts, should be ordered by ascending part number"
+          items:
+            $ref: "#/components/schemas/UploadPart"
+      required:
+        - parts
+```
+
+### Support and discover
+
+Presign support is a capability lakectl discover before switching to use presign for upload or download from lakeFS.
+The multipart upload support will be part of the storage capability add `pre_sign_support_parts` optional field that when set to `true` the user can perform multipart upload using the new API.
+
+```yaml
+    StorageConfig:
+      type: object
+      required:
+        - blockstore_type
+        - blockstore_namespace_example
+        - blockstore_namespace_ValidityRegex
+        - pre_sign_support
+        - pre_sign_support_ui
+        - import_support
+        - import_validity_regex
+      properties:
+        blockstore_type:
+          type: string
+        blockstore_namespace_example:
+          type: string
+        blockstore_namespace_ValidityRegex:
+          type: string
+        default_namespace_prefix:
+          type: string
+        pre_sign_support:
+          type: boolean
+        pre_sign_support_ui:
+          type: boolean
+        import_support:
+          type: boolean
+        import_validity_regex:
+          type: string
+        pre_sign_support_parts:
```

Comment:
If the presigned requirement is a limitation for the purposes of an MVP, then it should be a capability of the feature, as it is in other endpoints.

I'd suggest naming the config property something like `multipart_upload_support`. Especially since we already have `pre_sign_support`. The current limitation is that you can't set `multipart_upload_support` to `true` if `pre_sign_support` is set to `false`. We can "enforce" it this way.  

If we decide to allow multipart uploads without the presign requirement, we can remove this check.

Reply:
Update, the client will have to check both capabilities presigned and mutlipart supprort before using the API. The curent implementation is not return true for multipart upload support if presigned is off, as it is the only configuration that works.